src/js/form-model.js

```diff
@@ -1561,30 +1561,18 @@ FormModel.prototype.evaluate = function (
     }
 
     // try native to see if that works... (will not work if the expr contains custom OpenRosa functions)
-    if (
-        tryNative &&
```

Comment:
Note, it is unfortunately not safe to always try native because in this ecosystem we do things that can be evaluated by a native evaluator but return an incorrect result for ODK XForms. I'm surprised no tests failed.

The main (and perhaps only) things are comparisons and arithmetic with date/dateTime strings.

Reply:
Oh, this is a shame. I see a few other cases called out in the [spec](https://getodk.github.io/xforms-spec/#xpath-functions). This feels like something we could maybe revisit with the [tree-sitter-xpath](https://github.com/eyelidlessness/tree-sitter-xpath) grammar, which is fast enough that we'd still see significant performance improvements.

Reply:
Yes, it's a huge shame. We've had discussion about a turbo mode that requires form designer to wrap such strings with `date` and `date-time`.

Yes, indeed, good find. There are some native XPath 1.0. functions that have deviating behavior so some of those would also be an issue. 

Curious about tree-sitter! Will it be able to find out if the value of /path/to/node is a date string?

Reply:
Alone, the tree-sitter grammar won't be able to do any kind of static type analysis, that's not available in the AST. Here's ways I'd imagine it could help for this case:

- Rule out expressions with functions we know deviate (although we can skip some because they'll fail to compile, which should perform similarly).
- Rule out expressions where we know an argument position is of `date`/`date-time` type.
- Relate nodeset expressions to their bindings, to identify their type and rule out expressions with those nodesets[^1].
- Rule out expressions with literals which would be treated as dates.

If all of this sounds like it has overlap with openrosa-xpath-evaluator's parsing responsibilities... it does, heh. But it would probably be a good fit for this case, because tree-sitter is _exceptionally fast_.

[^1]: This dovetails with other prototype work I've explored, identifying nodeset subexpressions to determine their dependencies. This already works for a huge set of expressions I pulled from openrosa-xpath-evaluator's tests, you can see the [test fixtures used to validating that here](https://github.com/eyelidlessness/tree-sitter-xpath/blob/main/src/orxe-paths.ts). The grammar has proven pretty reliable so far. You can see example usage [here](https://github.com/eyelidlessness/tree-sitter-xpath/blob/main/src/grammar.test.ts#L44-L52) and [here](https://github.com/eyelidlessness/tree-sitter-xpath/blob/main/src/grammar.test.ts#L87-L98) to find nodeset sub-expressions. I have additional prototype work (currently only local) for resolving those sub-expressions to actual related nodes, which so far has worked for everything I've tried except with relative nested-subexpressions (e.g. in a predicate).

Reply:
Cool stuff! 

> I have additional prototype work (currently only local) for resolving those sub-expressions to actual related nodes, which so far has worked for everything

I'm guessing the challenge might be to prevent this from becoming too costly (possibly negate performance improvements achieved by sending to the native evaluator). Will be awesome if that is possible!

Reply:
It’s more than possible, it’s a reality! I’ll push up instructions for running and measuring the subexpression logic when I get a chance, but for now I’ll just say that finding subexpressions in all of the cases I pulled from the current evaluator test suite averages 1ms or less even when my computer is throttling under heavy load. In my local stress testing, the native evaluator is also generally ~1ms, the extended evaluator is generally 10+ms.